app/drivers/kscan/debounce.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2021 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#pragma once
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <sys/util.h>
+
+#define DEBOUNCE_COUNTER_BITS 14
+#define DEBOUNCE_COUNTER_MAX BIT_MASK(DEBOUNCE_COUNTER_BITS)
+
+struct debounce_state {
+    bool pressed : 1;
+    bool changed : 1;
+    uint16_t counter : DEBOUNCE_COUNTER_BITS;
+};
+
+struct debounce_config {
+    /** Duration a switch must be pressed to latch as pressed. */
+    uint32_t debounce_press_ms;
+    /** Duration a switch must be released to latch as released. */
+    uint32_t debounce_release_ms;
+};
+
+/**
+ * Debounces one switch.
+ *
+ * @note For the timings in "config" to be accurate, this currently assumes that
```

Comment:
Don't love this gotcha. Can we pass a duration to this function as well for "time elapsed"?

Reply:
Yes. I assume we would want to be able to scan faster than 1000 Hz, since having something capable of running Zephyr but not capable of scanning at 1000 Hz seems unlikely?

If we want that, then we can make the counter be a fixed point number of milliseconds, but then we need to decide how many bits to use for the fractional component. Assuming we kept the 14-bit counter, some reasonable splits:

Integer bits | Fractional bits | Max debounce time (ms) | Min timer resolution (ms)
------------|------------------|------|-----| 
14 (current) | 0 | 16,383 | 1 |
10 | 4 | 1023 | 0.0625
9 | 5 | 511 | 0.0313
8 | 6 | 255 | 0.0156
7 | 7 | 127 | 0.0078
6 | 8 | 63 | 0.0039
5 | 9 | 31 | 0.0020

Then the scanning routine needs to measure how long it's been in the appropriate fraction of a millisecond instead of just counting scans. This could lead to rounding errors depending on your scan rate though. 

For an exaggerated example, say you chose to scan at 3000 Hz with a minimum timer resolution of 1/16 ms and a debounce period of 1 ms. You would expect it to take 3 scans to debounce a key press since each scan would be 1/3 ms apart, but that rounds down to 5/16. After three scans the counter is now 15/16 ms, so it actually would take 4 scans. 

Maybe this is all too minor to matter, but the reason I did it the way it is now is because I didn't want to have to deal with that. Any suggestions on how to do this better are welcome.